core/extraction/towns_fund.py

```diff
@@ -0,0 +1,454 @@
+    """
+    Extract data from Towns Fund Reporting Template into column headed Pandas DataFrames.
+
+    :param df_ingest: DataFrame of parsed Excel data.
+    :return: Dictionary of extracted "tables" as DataFrames.
+    """
+
+    # TODO: check data for all DM v3.7 tabs is now extracted
+    towns_fund_extracted = {"df_programme_extracted": extract_programme(df_ingest["2 - Project Admin"])}
+    towns_fund_extracted["df_projects_extracted"] = extract_project(df_ingest["2 - Project Admin"])
+    number_of_projects = len(towns_fund_extracted["df_projects_extracted"].index)
+    towns_fund_extracted["df_programme_progress_extracted"] = extract_programme_progress(
+        df_ingest["3 - Programme Progress"]
+    )
+    towns_fund_extracted["df_project_progress_extracted"] = extract_project_progress(
+        df_ingest["3 - Programme Progress"]
+    )
+    towns_fund_extracted["df_funding_questions_extracted"] = extract_funding_questions(
+        df_ingest["4a - Funding Profiles"]
+    )
+    towns_fund_extracted["df_funding_comments_extracted"] = extract_funding_comments(
+        df_ingest["4a - Funding Profiles"], number_of_projects
+    )
+    towns_fund_extracted["df_funding_extracted"] = extract_funding_data(
+        df_ingest["4a - Funding Profiles"], number_of_projects
+    )
+
+    towns_fund_extracted["df_psi_extracted"] = extract_psi(df_ingest["4b - PSI"])
+
+    towns_fund_extracted["df_outputs_extracted"] = extract_outputs(df_ingest["5 - Project Outputs"], number_of_projects)
+    towns_fund_extracted["df_outcomes_extracted"] = extract_outcomes(df_ingest["6 - Outcomes"])
+
+    # separated from "outcomes" as these are in a different format, with greater date period granularity
+    towns_fund_extracted["df_outcomes_footfall_extracted"] = extract_footfall_outcomes(df_ingest["6 - Outcomes"])
+    towns_fund_extracted["df_programme_risks_extracted"] = extract_programme_risks(df_ingest["7 - Risk Register"])
+
+    # risks: cancelled projects show up, with nan cells in their section.
+    towns_fund_extracted["df_project_risks_extracted"] = extract_project_risks(
+        df_ingest["7 - Risk Register"], number_of_projects
+    )
+
+    return towns_fund_extracted
+
+
+def extract_programme(df_programme: pd.DataFrame) -> pd.DataFrame:
+    """
+    Extract package information from a DataFrame.
+
+    Input dataframe is parsed from Excel spreadsheet: "Towns Fund reporting template".
+    Specifically Project work sheet, parsed as dataframe.
+
+    :param df_programme: Input DataFrame containing data.
+    :return: Extracted DataFrame containing programme data.
+    """
+
+    # strip out everything other than "SECTION A..." in spreadsheet.
+    df_programme = df_programme.iloc[5:20, 2:5]
+
+    # rename col headers for ease
+    df_programme.columns = [0, 1, 2]
+
+    # combine first 2 cols into 1, filling in blank (merged) cells
+    field_names_first = [x := y if y is not np.nan else x for y in df_programme[0]]  # noqa: F841,F821
```

Comment:
How come you're using the `:=` notation here? I can't see how it's being used but I might be missing something?

Reply:
ahh very snazzy!

Reply:
the only minor downer on this is the x also persists in namespace after the comprehension has finished... :D